{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of the Chronic lymphocytic leukemia (CLL) \n",
    "\n",
    "\n",
    "The CLL RNA-Seq data was obtained from a study published by Ferreira et al. in 2014 [1]. This study included a deep RNA sequencing in different subpopulations of B-lymphocytes from healthy individuals and CLL cells from a cohort of 98 patients and characterized the CLL transcriptional landscape with unprecedented resolution. The transcriptomic architecture of CLL uncovered here refines the biological characterization of the disease and opens new perspectives for the clinical management of patients. \n",
    "\n",
    "The data is deposited on the European Genome-Phenome Archive (EGA, http://www.ebi.ac.uk/ega) under the study ID EGAS00001000374. The study is comprised of 98 CLL samples and 9 normal samples (each normal samples has three replicates).  \n",
    "\n",
    "We selected 97 CLL samples from the dataset EGAD00001000258. One of the samples (EGAR00001072814) was not labbeled with the mutation state, therefore we decided not to use this sample in our study.  \n",
    "\n",
    "We included in this repository a TSV file with the name of the samples and its classification. The access to the raw data should be requested to the EGA archive. The raw file names should be transformed to the one used in this study to use this set of scripts and code.\n",
    "\n",
    "### References\n",
    "\n",
    "1.\tFerreira PG, Jares P, Rico D, Gómez-López G, Martínez-Trillos A, Villamor N, Ecker S, González-Pérez A, Knowles DG, Monlong J et al: Transcriptome characterization by RNA sequencing identifies a major molecular and clinical subdivision in chronic lymphocytic leukemia. Genome Res 2014, 24(2):212-226.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the environment variables to be used during the project \n",
    "\n",
    "import os\n",
    "import re\n",
    "import pandas\n",
    "import math\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from six.moves import cPickle as pickle\n",
    "from pandas import read_csv\n",
    "\n",
    "###############################################################\n",
    "#\n",
    "#    Update this path\n",
    "#\n",
    "###############################################################\n",
    "os.environ['WORKDIR'] = os.path.abspath('/panfs/pan1.be-md.ncbi.nlm.nih.gov/alt_splicing/bioNotebook-DiffExpIR')\n",
    "###############################################################\n",
    "\n",
    "os.environ['CONFIG'] = os.environ['WORKDIR'] + '/config'\n",
    "os.environ['DATA'] = os.environ['WORKDIR'] + '/data'\n",
    "os.environ['BIN'] = os.environ['WORKDIR'] + '/bin'\n",
    "os.environ['RESULTS'] = os.environ['WORKDIR'] + '/results'\n",
    "os.environ['DOC'] = os.environ['WORKDIR'] + '/docs'\n",
    "os.environ['SRC'] = os.environ['WORKDIR'] + '/src'\n",
    "\n",
    "os.environ['DATADIR'] = 'CLL_EGAS00001000374'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating result folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd ${RESULTS}\n",
    "if [ ! -e ${DATADIR} ]\n",
    "then\n",
    "    mkdir ${DATADIR}\n",
    "fi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Submitting the alignments to the SGE cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your job 6421672 (\"EGAR00001072723_CLL250909_7\") has been submitted\n",
      "Your job 6421673 (\"EGAR00001072724_CLL280809_7\") has been submitted\n",
      "Your job 6421674 (\"EGAR00001072725_C0EWEACXX_3_19\") has been submitted\n",
      "Your job 6421675 (\"EGAR00001072726_D030UACXX_7_1\") has been submitted\n",
      "Your job 6421677 (\"EGAR00001072727_62JP4AAXX_1\") has been submitted\n",
      "Your job 6421678 (\"EGAR00001072728_D0933ACXX_2_2\") has been submitted\n",
      "Your job 6421679 (\"EGAR00001072729_62JP4AAXX_8\") has been submitted\n",
      "Your job 6421680 (\"EGAR00001072730_626WGAAXX_1\") has been submitted\n",
      "Your job 6421682 (\"EGAR00001072731_626WGAAXX_2\") has been submitted\n",
      "Your job 6421683 (\"EGAR00001072732_626WGAAXX_3\") has been submitted\n",
      "Your job 6421684 (\"EGAR00001072733_626WGAAXX_4\") has been submitted\n",
      "Your job 6421685 (\"EGAR00001072734_62JP4AAXX_2\") has been submitted\n",
      "Your job 6421687 (\"EGAR00001072735_626WGAAXX_7\") has been submitted\n",
      "Your job 6421689 (\"EGAR00001072736_626WGAAXX_8\") has been submitted\n",
      "Your job 6421690 (\"EGAR00001072737_627LPAAXX_1\") has been submitted\n",
      "Your job 6421691 (\"EGAR00001072738_627LPAAXX_2\") has been submitted\n",
      "Your job 6421693 (\"EGAR00001072739_D0933ACXX_1_3\") has been submitted\n",
      "Your job 6421694 (\"EGAR00001072740_627LPAAXX_3\") has been submitted\n",
      "Your job 6421696 (\"EGAR00001072741_627LPAAXX_4\") has been submitted\n",
      "Your job 6421697 (\"EGAR00001072742_627LPAAXX_7\") has been submitted\n",
      "Your job 6421699 (\"EGAR00001072743_D0933ACXX_2_4\") has been submitted\n",
      "Your job 6421700 (\"EGAR00001072744_627LPAAXX_8\") has been submitted\n",
      "Your job 6421701 (\"EGAR00001072745_D0933ACXX_3_5\") has been submitted\n",
      "Your job 6421705 (\"EGAR00001072746_627M8AAXX_1\") has been submitted\n",
      "Your job 6421707 (\"EGAR00001072747_627M8AAXX_2\") has been submitted\n",
      "Your job 6421708 (\"EGAR00001072748_D030UACXX_7_2\") has been submitted\n",
      "Your job 6421710 (\"EGAR00001072749_D030UACXX_7_3\") has been submitted\n",
      "Your job 6421712 (\"EGAR00001072750_627M8AAXX_3\") has been submitted\n",
      "Your job 6421714 (\"EGAR00001072751_D030UACXX_7_4\") has been submitted\n",
      "Your job 6421715 (\"EGAR00001072752_626WAAAXX_1\") has been submitted\n",
      "Your job 6421716 (\"EGAR00001072753_D0933ACXX_3_6\") has been submitted\n",
      "Your job 6421717 (\"EGAR00001072754_626WAAAXX_2\") has been submitted\n",
      "Your job 6421719 (\"EGAR00001072755_D030UACXX_8_5\") has been submitted\n",
      "Your job 6421721 (\"EGAR00001072756_D0933ACXX_3_7\") has been submitted\n",
      "Your job 6421722 (\"EGAR00001072757_D0933ACXX_5_8\") has been submitted\n",
      "Your job 6421723 (\"EGAR00001072758_D030UACXX_8_6\") has been submitted\n",
      "Your job 6421725 (\"EGAR00001072759_D030UACXX_8_7\") has been submitted\n",
      "Your job 6421726 (\"EGAR00001072760_D030UACXX_8_8\") has been submitted\n",
      "Your job 6421727 (\"EGAR00001072761_D0933ACXX_5_9\") has been submitted\n",
      "Your job 6421729 (\"EGAR00001072762_B02VRACXX_5_9\") has been submitted\n",
      "Your job 6421730 (\"EGAR00001072763_B02VRACXX_5_10\") has been submitted\n",
      "Your job 6421732 (\"EGAR00001072764_626WAAAXX_3\") has been submitted\n",
      "Your job 6421733 (\"EGAR00001072765_626WAAAXX_4\") has been submitted\n",
      "Your job 6421736 (\"EGAR00001072766_B02VRACXX_5_11\") has been submitted\n",
      "Your job 6421737 (\"EGAR00001072767_B02VRACXX_5_12\") has been submitted\n",
      "Your job 6421739 (\"EGAR00001072768_B02VRACXX_6_9\") has been submitted\n",
      "Your job 6421741 (\"EGAR00001072769_B02VRACXX_6_10\") has been submitted\n",
      "Your job 6421742 (\"EGAR00001072770_B02VRACXX_6_11\") has been submitted\n",
      "Your job 6421743 (\"EGAR00001072771_B02VRACXX_6_12\") has been submitted\n",
      "Your job 6421744 (\"EGAR00001072772_B02VRACXX_7_1\") has been submitted\n",
      "Your job 6421747 (\"EGAR00001072773_B02VRACXX_7_2\") has been submitted\n",
      "Your job 6421750 (\"EGAR00001072774_B02VRACXX_7_3\") has been submitted\n",
      "Your job 6421751 (\"EGAR00001072775_B02VRACXX_7_4\") has been submitted\n",
      "Your job 6421756 (\"EGAR00001072776_B02VRACXX_8_5\") has been submitted\n",
      "Your job 6421760 (\"EGAR00001072777_D0933ACXX_5_10\") has been submitted\n",
      "Your job 6421763 (\"EGAR00001072778_D0933ACXX_6_11\") has been submitted\n",
      "Your job 6421765 (\"EGAR00001072779_B02VRACXX_8_6\") has been submitted\n",
      "Your job 6421767 (\"EGAR00001072780_B02VRACXX_8_7\") has been submitted\n",
      "Your job 6421771 (\"EGAR00001072781_D0933ACXX_7_12\") has been submitted\n",
      "Your job 6421776 (\"EGAR00001072782_D0933ACXX_7_1\") has been submitted\n",
      "Your job 6421778 (\"EGAR00001072783_D0933ACXX_7_2\") has been submitted\n",
      "Your job 6421781 (\"EGAR00001072784_D0933ACXX_8_3\") has been submitted\n",
      "Your job 6421784 (\"EGAR00001072785_D0933ACXX_8_4\") has been submitted\n",
      "Your job 6421786 (\"EGAR00001072786_D0933ACXX_8_5\") has been submitted\n",
      "Your job 6421788 (\"EGAR00001072787_B02VRACXX_8_8\") has been submitted\n",
      "Your job 6421791 (\"EGAR00001072788_D0933ACXX_3_1\") has been submitted\n",
      "Your job 6421792 (\"EGAR00001072789_D0933ACXX_5_2\") has been submitted\n",
      "Your job 6421793 (\"EGAR00001072790_D0933ACXX_2_3\") has been submitted\n",
      "Your job 6421795 (\"EGAR00001072791_D0933ACXX_1_4\") has been submitted\n",
      "Your job 6421796 (\"EGAR00001072792_D0933ACXX_6_5\") has been submitted\n",
      "Your job 6421797 (\"EGAR00001072793_D0933ACXX_6_6\") has been submitted\n",
      "Your job 6421798 (\"EGAR00001072794_D0933ACXX_6_7\") has been submitted\n",
      "Your job 6421801 (\"EGAR00001072795_D0933ACXX_7_8\") has been submitted\n",
      "Your job 6421802 (\"EGAR00001072796_D0933ACXX_1_1\") has been submitted\n",
      "Your job 6421804 (\"EGAR00001072797_D0933ACXX_8_9\") has been submitted\n",
      "Your job 6421805 (\"EGAR00001072798_C01CLACXX_7_10\") has been submitted\n",
      "Your job 6421806 (\"EGAR00001072799_C01CLACXX_7_11\") has been submitted\n",
      "Your job 6421808 (\"EGAR00001072800_C01CLACXX_7_12\") has been submitted\n",
      "Your job 6421811 (\"EGAR00001072801_C01CLACXX_7_1\") has been submitted\n",
      "Your job 6421814 (\"EGAR00001072802_C01CLACXX_8_2\") has been submitted\n",
      "Your job 6421816 (\"EGAR00001072803_C01CLACXX_8_3\") has been submitted\n",
      "Your job 6421817 (\"EGAR00001072804_C01CLACXX_8_4\") has been submitted\n",
      "Your job 6421821 (\"EGAR00001072805_C01CLACXX_8_1\") has been submitted\n",
      "Your job 6421822 (\"EGAR00001072806_C01BRACXX_4_2\") has been submitted\n",
      "Your job 6421824 (\"EGAR00001072807_C01BRACXX_4_3\") has been submitted\n",
      "Your job 6421825 (\"EGAR00001072808_C01BRACXX_5_4\") has been submitted\n",
      "Your job 6421826 (\"EGAR00001072809_C01BRACXX_5_5\") has been submitted\n",
      "Your job 6421829 (\"EGAR00001072810_C01BRACXX_5_6\") has been submitted\n",
      "Your job 6421830 (\"EGAR00001072811_C01BRACXX_6_7\") has been submitted\n",
      "Your job 6421831 (\"EGAR00001072812_C01BRACXX_6_8\") has been submitted\n",
      "Your job 6421832 (\"EGAR00001072813_C01BRACXX_6_9\") has been submitted\n",
      "Your job 6421834 (\"EGAR00001072815_C01BRACXX_7_11\") has been submitted\n",
      "Your job 6421836 (\"EGAR00001072816_D0933ACXX_1_2\") has been submitted\n",
      "Your job 6421837 (\"EGAR00001072817_C01BRACXX_7_12\") has been submitted\n",
      "Your job 6421838 (\"EGAR00001072818_C01BRACXX_8_1\") has been submitted\n",
      "Your job 6421840 (\"EGAR00001072819_C01BRACXX_8_2\") has been submitted\n",
      "Your job 6421841 (\"EGAR00001072820_C01BRACXX_8_3\") has been submitted\n",
      "Your job 6421843 (\"EGAR00001072821_C05MPACXX_4_16\") has been submitted\n",
      "Your job 6421844 (\"EGAR00001072822_C0APYACXX_4_16\") has been submitted\n",
      "Your job 6421846 (\"EGAR00001072823_D0KJ7ACXX_2_16\") has been submitted\n",
      "Your job 6421849 (\"EGAR00001072824_C05MPACXX_4_18\") has been submitted\n",
      "Your job 6421851 (\"EGAR00001072825_C0APYACXX_4_18\") has been submitted\n",
      "Your job 6421853 (\"EGAR00001072826_D0KJ7ACXX_2_18\") has been submitted\n",
      "Your job 6421855 (\"EGAR00001072827_C05MPACXX_4_19\") has been submitted\n",
      "Your job 6421856 (\"EGAR00001072828_C0APYACXX_4_19\") has been submitted\n",
      "Your job 6421858 (\"EGAR00001072829_D0KJ7ACXX_2_19\") has been submitted\n",
      "Your job 6421859 (\"EGAR00001072830_C05MPACXX_3_10\") has been submitted\n",
      "Your job 6421860 (\"EGAR00001072831_C0APYACXX_3_10\") has been submitted\n",
      "Your job 6421861 (\"EGAR00001072832_D0KJ7ACXX_1_10\") has been submitted\n",
      "Your job 6421862 (\"EGAR00001072833_C05MPACXX_3_11\") has been submitted\n",
      "Your job 6421863 (\"EGAR00001072834_C0APYACXX_3_11\") has been submitted\n",
      "Your job 6421864 (\"EGAR00001072835_D0KJ7ACXX_1_11\") has been submitted\n",
      "Your job 6421866 (\"EGAR00001072836_C05MPACXX_3_12\") has been submitted\n",
      "Your job 6421867 (\"EGAR00001072837_C0APYACXX_3_12\") has been submitted\n",
      "Your job 6421868 (\"EGAR00001072838_D0KJ7ACXX_1_12\") has been submitted\n",
      "Your job 6421869 (\"EGAR00001072839_C05MPACXX_3_13\") has been submitted\n",
      "Your job 6421872 (\"EGAR00001072840_C0APYACXX_3_13\") has been submitted\n",
      "Your job 6421873 (\"EGAR00001072841_D0KJ7ACXX_1_13\") has been submitted\n",
      "Your job 6421874 (\"EGAR00001072842_C05MPACXX_3_14\") has been submitted\n",
      "Your job 6421876 (\"EGAR00001072843_C0APYACXX_3_14\") has been submitted\n",
      "Your job 6421877 (\"EGAR00001072844_D0KJ7ACXX_1_14\") has been submitted\n",
      "Your job 6421878 (\"EGAR00001072845_C05MPACXX_4_15\") has been submitted\n",
      "Your job 6421880 (\"EGAR00001072846_C0APYACXX_4_15\") has been submitted\n",
      "Your job 6421881 (\"EGAR00001072847_D0KJ7ACXX_2_15\") has been submitted\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ${RESULTS}/${DATADIR}\n",
    "if [ ! -e \"STAR_hg19_Human\" ]\n",
    "then\n",
    "    mkdir STAR_hg19_Human\n",
    "    cd STAR_hg19_Human\n",
    "    for f in ${DATA}/${DATADIR}/*_1.fastq.gz\n",
    "    do\n",
    "        n=`basename $f`\n",
    "        n=${n%_1.fastq.gz}\n",
    "        qsub -v SGE_FACILITIES -l h_rt=86400,h_vmem=32G -N ${n} -o ${n}.out -e ${n}.err -pe multicore 16 -m n $BIN/star.sh ${n} $DATA/${DATADIR} $DATA/hg19/Sequence/STAR\n",
    "    done\n",
    "fi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Submitting the TPMCalculator to the SGE cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your job 6424335 (\"EGAR00001072723_CLL250909_7\") has been submitted\n",
      "Your job 6424337 (\"EGAR00001072724_CLL280809_7\") has been submitted\n",
      "Your job 6424339 (\"EGAR00001072725_C0EWEACXX_3_19\") has been submitted\n",
      "Your job 6424341 (\"EGAR00001072726_D030UACXX_7_1\") has been submitted\n",
      "Your job 6424342 (\"EGAR00001072727_62JP4AAXX_1\") has been submitted\n",
      "Your job 6424343 (\"EGAR00001072728_D0933ACXX_2_2\") has been submitted\n",
      "Your job 6424345 (\"EGAR00001072729_62JP4AAXX_8\") has been submitted\n",
      "Your job 6424346 (\"EGAR00001072730_626WGAAXX_1\") has been submitted\n",
      "Your job 6424347 (\"EGAR00001072731_626WGAAXX_2\") has been submitted\n",
      "Your job 6424349 (\"EGAR00001072732_626WGAAXX_3\") has been submitted\n",
      "Your job 6424350 (\"EGAR00001072733_626WGAAXX_4\") has been submitted\n",
      "Your job 6424351 (\"EGAR00001072734_62JP4AAXX_2\") has been submitted\n",
      "Your job 6424353 (\"EGAR00001072735_626WGAAXX_7\") has been submitted\n",
      "Your job 6424354 (\"EGAR00001072736_626WGAAXX_8\") has been submitted\n",
      "Your job 6424355 (\"EGAR00001072737_627LPAAXX_1\") has been submitted\n",
      "Your job 6424357 (\"EGAR00001072738_627LPAAXX_2\") has been submitted\n",
      "Your job 6424358 (\"EGAR00001072739_D0933ACXX_1_3\") has been submitted\n",
      "Your job 6424359 (\"EGAR00001072740_627LPAAXX_3\") has been submitted\n",
      "Your job 6424361 (\"EGAR00001072741_627LPAAXX_4\") has been submitted\n",
      "Your job 6424362 (\"EGAR00001072742_627LPAAXX_7\") has been submitted\n",
      "Your job 6424364 (\"EGAR00001072743_D0933ACXX_2_4\") has been submitted\n",
      "Your job 6424366 (\"EGAR00001072744_627LPAAXX_8\") has been submitted\n",
      "Your job 6424368 (\"EGAR00001072745_D0933ACXX_3_5\") has been submitted\n",
      "Your job 6424371 (\"EGAR00001072746_627M8AAXX_1\") has been submitted\n",
      "Your job 6424372 (\"EGAR00001072747_627M8AAXX_2\") has been submitted\n",
      "Your job 6424374 (\"EGAR00001072748_D030UACXX_7_2\") has been submitted\n",
      "Your job 6424377 (\"EGAR00001072749_D030UACXX_7_3\") has been submitted\n",
      "Your job 6424378 (\"EGAR00001072750_627M8AAXX_3\") has been submitted\n",
      "Your job 6424380 (\"EGAR00001072751_D030UACXX_7_4\") has been submitted\n",
      "Your job 6424382 (\"EGAR00001072752_626WAAAXX_1\") has been submitted\n",
      "Your job 6424384 (\"EGAR00001072753_D0933ACXX_3_6\") has been submitted\n",
      "Your job 6424386 (\"EGAR00001072754_626WAAAXX_2\") has been submitted\n",
      "Your job 6424387 (\"EGAR00001072755_D030UACXX_8_5\") has been submitted\n",
      "Your job 6424389 (\"EGAR00001072756_D0933ACXX_3_7\") has been submitted\n",
      "Your job 6424390 (\"EGAR00001072757_D0933ACXX_5_8\") has been submitted\n",
      "Your job 6424393 (\"EGAR00001072758_D030UACXX_8_6\") has been submitted\n",
      "Your job 6424395 (\"EGAR00001072759_D030UACXX_8_7\") has been submitted\n",
      "Your job 6424397 (\"EGAR00001072760_D030UACXX_8_8\") has been submitted\n",
      "Your job 6424399 (\"EGAR00001072761_D0933ACXX_5_9\") has been submitted\n",
      "Your job 6424402 (\"EGAR00001072762_B02VRACXX_5_9\") has been submitted\n",
      "Your job 6424404 (\"EGAR00001072763_B02VRACXX_5_10\") has been submitted\n",
      "Your job 6424406 (\"EGAR00001072764_626WAAAXX_3\") has been submitted\n",
      "Your job 6424408 (\"EGAR00001072765_626WAAAXX_4\") has been submitted\n",
      "Your job 6424410 (\"EGAR00001072766_B02VRACXX_5_11\") has been submitted\n",
      "Your job 6424411 (\"EGAR00001072767_B02VRACXX_5_12\") has been submitted\n",
      "Your job 6424415 (\"EGAR00001072768_B02VRACXX_6_9\") has been submitted\n",
      "Your job 6424417 (\"EGAR00001072769_B02VRACXX_6_10\") has been submitted\n",
      "Your job 6424421 (\"EGAR00001072770_B02VRACXX_6_11\") has been submitted\n",
      "Your job 6424422 (\"EGAR00001072771_B02VRACXX_6_12\") has been submitted\n",
      "Your job 6424424 (\"EGAR00001072772_B02VRACXX_7_1\") has been submitted\n",
      "Your job 6424427 (\"EGAR00001072773_B02VRACXX_7_2\") has been submitted\n",
      "Your job 6424428 (\"EGAR00001072774_B02VRACXX_7_3\") has been submitted\n",
      "Your job 6424431 (\"EGAR00001072775_B02VRACXX_7_4\") has been submitted\n",
      "Your job 6424434 (\"EGAR00001072776_B02VRACXX_8_5\") has been submitted\n",
      "Your job 6424437 (\"EGAR00001072777_D0933ACXX_5_10\") has been submitted\n",
      "Your job 6424440 (\"EGAR00001072778_D0933ACXX_6_11\") has been submitted\n",
      "Your job 6424441 (\"EGAR00001072779_B02VRACXX_8_6\") has been submitted\n",
      "Your job 6424443 (\"EGAR00001072780_B02VRACXX_8_7\") has been submitted\n",
      "Your job 6424447 (\"EGAR00001072781_D0933ACXX_7_12\") has been submitted\n",
      "Your job 6424449 (\"EGAR00001072782_D0933ACXX_7_1\") has been submitted\n",
      "Your job 6424451 (\"EGAR00001072783_D0933ACXX_7_2\") has been submitted\n",
      "Your job 6424455 (\"EGAR00001072784_D0933ACXX_8_3\") has been submitted\n",
      "Your job 6424456 (\"EGAR00001072785_D0933ACXX_8_4\") has been submitted\n",
      "Your job 6424459 (\"EGAR00001072786_D0933ACXX_8_5\") has been submitted\n",
      "Your job 6424462 (\"EGAR00001072787_B02VRACXX_8_8\") has been submitted\n",
      "Your job 6424468 (\"EGAR00001072788_D0933ACXX_3_1\") has been submitted\n",
      "Your job 6424470 (\"EGAR00001072789_D0933ACXX_5_2\") has been submitted\n",
      "Your job 6424471 (\"EGAR00001072790_D0933ACXX_2_3\") has been submitted\n",
      "Your job 6424472 (\"EGAR00001072791_D0933ACXX_1_4\") has been submitted\n",
      "Your job 6424474 (\"EGAR00001072792_D0933ACXX_6_5\") has been submitted\n",
      "Your job 6424476 (\"EGAR00001072793_D0933ACXX_6_6\") has been submitted\n",
      "Your job 6424477 (\"EGAR00001072794_D0933ACXX_6_7\") has been submitted\n",
      "Your job 6424479 (\"EGAR00001072795_D0933ACXX_7_8\") has been submitted\n",
      "Your job 6424480 (\"EGAR00001072796_D0933ACXX_1_1\") has been submitted\n",
      "Your job 6424482 (\"EGAR00001072797_D0933ACXX_8_9\") has been submitted\n",
      "Your job 6424485 (\"EGAR00001072798_C01CLACXX_7_10\") has been submitted\n",
      "Your job 6424488 (\"EGAR00001072799_C01CLACXX_7_11\") has been submitted\n",
      "Your job 6424490 (\"EGAR00001072800_C01CLACXX_7_12\") has been submitted\n",
      "Your job 6424503 (\"EGAR00001072801_C01CLACXX_7_1\") has been submitted\n",
      "Your job 6424509 (\"EGAR00001072802_C01CLACXX_8_2\") has been submitted\n",
      "Your job 6424518 (\"EGAR00001072803_C01CLACXX_8_3\") has been submitted\n",
      "Your job 6424521 (\"EGAR00001072804_C01CLACXX_8_4\") has been submitted\n",
      "Your job 6424522 (\"EGAR00001072805_C01CLACXX_8_1\") has been submitted\n",
      "Your job 6424526 (\"EGAR00001072806_C01BRACXX_4_2\") has been submitted\n",
      "Your job 6424527 (\"EGAR00001072807_C01BRACXX_4_3\") has been submitted\n",
      "Your job 6424528 (\"EGAR00001072808_C01BRACXX_5_4\") has been submitted\n",
      "Your job 6424529 (\"EGAR00001072809_C01BRACXX_5_5\") has been submitted\n",
      "Your job 6424531 (\"EGAR00001072810_C01BRACXX_5_6\") has been submitted\n",
      "Your job 6424533 (\"EGAR00001072811_C01BRACXX_6_7\") has been submitted\n",
      "Your job 6424534 (\"EGAR00001072812_C01BRACXX_6_8\") has been submitted\n",
      "Your job 6424536 (\"EGAR00001072813_C01BRACXX_6_9\") has been submitted\n",
      "Your job 6424538 (\"EGAR00001072815_C01BRACXX_7_11\") has been submitted\n",
      "Your job 6424539 (\"EGAR00001072816_D0933ACXX_1_2\") has been submitted\n",
      "Your job 6424542 (\"EGAR00001072817_C01BRACXX_7_12\") has been submitted\n",
      "Your job 6424543 (\"EGAR00001072818_C01BRACXX_8_1\") has been submitted\n",
      "Your job 6424544 (\"EGAR00001072819_C01BRACXX_8_2\") has been submitted\n",
      "Your job 6424545 (\"EGAR00001072820_C01BRACXX_8_3\") has been submitted\n",
      "Your job 6424547 (\"EGAR00001072821_C05MPACXX_4_16\") has been submitted\n",
      "Your job 6424549 (\"EGAR00001072822_C0APYACXX_4_16\") has been submitted\n",
      "Your job 6424551 (\"EGAR00001072823_D0KJ7ACXX_2_16\") has been submitted\n",
      "Your job 6424553 (\"EGAR00001072824_C05MPACXX_4_18\") has been submitted\n",
      "Your job 6424554 (\"EGAR00001072825_C0APYACXX_4_18\") has been submitted\n",
      "Your job 6424555 (\"EGAR00001072826_D0KJ7ACXX_2_18\") has been submitted\n",
      "Your job 6424557 (\"EGAR00001072827_C05MPACXX_4_19\") has been submitted\n",
      "Your job 6424558 (\"EGAR00001072828_C0APYACXX_4_19\") has been submitted\n",
      "Your job 6424559 (\"EGAR00001072829_D0KJ7ACXX_2_19\") has been submitted\n",
      "Your job 6424560 (\"EGAR00001072830_C05MPACXX_3_10\") has been submitted\n",
      "Your job 6424562 (\"EGAR00001072831_C0APYACXX_3_10\") has been submitted\n",
      "Your job 6424563 (\"EGAR00001072832_D0KJ7ACXX_1_10\") has been submitted\n",
      "Your job 6424564 (\"EGAR00001072833_C05MPACXX_3_11\") has been submitted\n",
      "Your job 6424565 (\"EGAR00001072834_C0APYACXX_3_11\") has been submitted\n",
      "Your job 6424567 (\"EGAR00001072835_D0KJ7ACXX_1_11\") has been submitted\n",
      "Your job 6424569 (\"EGAR00001072836_C05MPACXX_3_12\") has been submitted\n",
      "Your job 6424571 (\"EGAR00001072837_C0APYACXX_3_12\") has been submitted\n",
      "Your job 6424572 (\"EGAR00001072838_D0KJ7ACXX_1_12\") has been submitted\n",
      "Your job 6424574 (\"EGAR00001072839_C05MPACXX_3_13\") has been submitted\n",
      "Your job 6424577 (\"EGAR00001072840_C0APYACXX_3_13\") has been submitted\n",
      "Your job 6424579 (\"EGAR00001072841_D0KJ7ACXX_1_13\") has been submitted\n",
      "Your job 6424580 (\"EGAR00001072842_C05MPACXX_3_14\") has been submitted\n",
      "Your job 6424581 (\"EGAR00001072843_C0APYACXX_3_14\") has been submitted\n",
      "Your job 6424583 (\"EGAR00001072844_D0KJ7ACXX_1_14\") has been submitted\n",
      "Your job 6424584 (\"EGAR00001072845_C05MPACXX_4_15\") has been submitted\n",
      "Your job 6424586 (\"EGAR00001072846_C0APYACXX_4_15\") has been submitted\n",
      "Your job 6424588 (\"EGAR00001072847_D0KJ7ACXX_2_15\") has been submitted\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ${RESULTS}/${DATADIR}\n",
    "if [ ! -e \"STAR_hg19_Human_TPM\" ]\n",
    "then\n",
    "    mkdir STAR_hg19_Human_TPM\n",
    "    cd STAR_hg19_Human_TPM\n",
    "    for f in ${RESULTS}/${DATADIR}/STAR_hg19_Human/*.bam\n",
    "    do\n",
    "        n=`basename $f`\n",
    "        qsub -v SGE_FACILITIES -l h_rt=86400,h_vmem=8G -N ${n%_sorted.bam} -o ${n%_sorted.bam}.out -e ${n%_sorted.bam}.err -m n -b y $BIN/TPMCalculator -g $DATA/hg19/Annotation/Genes/genes.gtf -b $f -p\n",
    "    done\n",
    "fi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading sample description data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ECGA_ID</th>\n",
       "      <th>Mut_Stat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGAR00001072723</td>\n",
       "      <td>UNM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EGAR00001072724</td>\n",
       "      <td>UNM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGAR00001072725</td>\n",
       "      <td>MUT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EGAR00001072726</td>\n",
       "      <td>MUT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EGAR00001072727</td>\n",
       "      <td>UNM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ECGA_ID Mut_Stat\n",
       "0  EGAR00001072723      UNM\n",
       "1  EGAR00001072724      UNM\n",
       "2  EGAR00001072725      MUT\n",
       "3  EGAR00001072726      MUT\n",
       "4  EGAR00001072727      UNM"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cll_samples = pandas.read_csv(os.path.join(os.environ['DATA'], os.environ['DATADIR'], 'cll_samples.tsv'), sep='\\t')\n",
    "cll_samples.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Python functions to process the data\n",
    "\n",
    "#### Function to parser the TPMCalculator data files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_data(ctrl_dir, ctl_samples, treat_dir, treat_samples):\n",
    "    samples = {}\n",
    "    for file in ctl_samples:\n",
    "        print(\"Parsing file %s\" % os.path.join(ctrl_dir, file))\n",
    "        data = pandas.read_csv(os.path.join(ctrl_dir, file), sep='\\t')\n",
    "        for chr in data.Chr.unique():\n",
    "            if '_' not in chr:\n",
    "                if chr not in samples:\n",
    "                    samples[chr] = {}\n",
    "                for g in data.loc[data['Chr'] == chr, 'Gene_Id'].unique():\n",
    "                    if g not in samples[chr]:\n",
    "                        samples[chr][g] = {}\n",
    "        for index, row in data.iterrows():\n",
    "            chr = row['Chr']\n",
    "            if '_' not in chr:\n",
    "                g = row['Gene_Id']\n",
    "                c = int(row['Type_Number'])\n",
    "                start = int(row['start']) + 1\n",
    "                end = int(row['end']) + 1\n",
    "                if c not in samples[chr][g]:\n",
    "                    samples[chr][g][c] = {'type': row['Type'], 'start': start, 'end': end,\n",
    "                                          'TPM_list_ctrl': [], 'Reads_list_ctrl': [], 'TPM_ctrl': 0,\n",
    "                                          'Reads_ctrl': 0,\n",
    "                                          'TPM_list_treat': [], 'Reads_list_treat': [], 'TPM_treat': 0,\n",
    "                                          'Reads_treat': 0}\n",
    "                tpm = float(row['TPM'])\n",
    "                if tpm < 10E-5:\n",
    "                    tpm = 10E-5\n",
    "                samples[chr][g][c]['TPM_list_ctrl'].append(tpm)\n",
    "                samples[chr][g][c]['Reads_list_ctrl'].append(int(row['Count_Reads']))\n",
    "                samples[chr][g][c]['TPM_ctrl'] = np.mean(samples[chr][g][c]['TPM_list_ctrl'])\n",
    "                samples[chr][g][c]['Reads_ctrl'] = np.mean(samples[chr][g][c]['Reads_list_ctrl'])\n",
    "    for file in treat_samples:\n",
    "        print(\"Parsing file %s\" % os.path.join(treat_dir, file))\n",
    "        data = pandas.read_csv(os.path.join(treat_dir, file), sep='\\t')\n",
    "        for chr in data.Chr.unique():\n",
    "            if '_' not in chr:\n",
    "                if chr not in samples:\n",
    "                    samples[chr] = {}\n",
    "                for g in data.loc[data['Chr'] == chr, 'Gene_Id'].unique():\n",
    "                    if g not in samples[chr]:\n",
    "                        samples[chr][g] = {}\n",
    "        for index, row in data.iterrows():\n",
    "            chr = row['Chr']\n",
    "            if '_' not in chr:\n",
    "                g = row['Gene_Id']\n",
    "                c = int(row['Type_Number'])\n",
    "                start = int(row['start']) + 1\n",
    "                end = int(row['end']) + 1\n",
    "                if c not in samples[chr][g]:\n",
    "                    samples[chr][g][c] = {'type': row['Type'], 'start': start, 'end': end,\n",
    "                                          'TPM_list_ctrl': [], 'Reads_list_ctrl': [], 'TPM_ctrl': 0,\n",
    "                                          'Reads_ctrl': 0,\n",
    "                                          'TPM_list_treat': [], 'Reads_list_treat': [], 'TPM_treat': 0,\n",
    "                                          'Reads_treat': 0}\n",
    "                tpm = float(row['TPM'])\n",
    "                if tpm < 10E-5:\n",
    "                    tpm = 10E-5\n",
    "                samples[chr][g][c]['TPM_list_treat'].append(tpm)\n",
    "                samples[chr][g][c]['Reads_list_treat'].append(int(row['Count_Reads']))\n",
    "                samples[chr][g][c]['TPM_treat'] = np.mean(samples[chr][g][c]['TPM_list_treat'])\n",
    "                samples[chr][g][c]['Reads_treat'] = np.mean(samples[chr][g][c]['Reads_list_treat'])\n",
    "    return samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating a list of samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(os.environ['RESULTS'] + '/' + os.environ['DATADIR'] + '/STAR_hg19_Human_TPM')\n",
    "samples = {'ctrl' : [], 'cll':[]}\n",
    "files = [name for root, dirs, files in os.walk('./') for name in files if name.endswith(\"_genes.ent\")]\n",
    "for f in files:\n",
    "    n = f.split('_')[0]\n",
    "    row = cll_samples.loc[cll_samples['ECGA_ID'] == n]\n",
    "    if row.values[0][1] == 'NBC':\n",
    "        samples['ctrl'].append(f)\n",
    "    else:\n",
    "        samples['cll'].append(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating a pickle file to store the Python objects used by the DiffExpIR algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(os.environ['RESULTS'] + '/' + os.environ['DATADIR'] + '/STAR_hg19_Human_TPM')\n",
    "pickle_file = 'samples_data.pickle'\n",
    "if not os.path.exists(pickle_file):    \n",
    "    try:\n",
    "        samples_data = extract_data('./', samples['ctrl'], './', samples['cll'])\n",
    "        print('Pickling ...')    \n",
    "        with open(pickle_file, 'wb') as f:\n",
    "            pickle.dump(samples_data, f, pickle.HIGHEST_PROTOCOL)\n",
    "    except Exception as e:\n",
    "        print('Unable to save data to', pickle_file, ':', e)\n",
    "else:\n",
    "    try:\n",
    "        with open(pickle_file, 'rb') as f:\n",
    "            samples_data = pickle.load(f)\n",
    "    except Exception as e:\n",
    "        print('Unable to process data from', pickle_file, ':', e)\n",
    "        raise"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Function to process one intron in a gene\n",
    "\n",
    "This function executes the T-Test function to calculate the P-Value associated to the comparison among the control and treated samples. \n",
    "\n",
    "Then, the r values, $r = log_{2}\\lgroup\\frac{TPM^{intron}_i}{(TPM^{exon}_{i-1} + TPM^{exon}_{i+1})}\\rgroup$, is calculated.\n",
    "\n",
    "This function returns for the intron: PValue, $-1.0 * log_{10}(PValue)$, $r_{ctrl}$, $r_{treat}$, $log_{2}\\lgroup\\frac{TPM^{intron}_{treat}}{TPM^{intron}_{ctrl}}\\rgroup$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_one(chr, g, a, samples):\n",
    "    equal_var = False\n",
    "    if len(samples[chr][g][a]['TPM_list_treat']) == len(samples[chr][g][a]['TPM_list_ctrl']):\n",
    "        equal_var = True\n",
    "    t, p = stats.ttest_ind(samples[chr][g][a]['TPM_list_treat'],\n",
    "                           samples[chr][g][a]['TPM_list_ctrl'], equal_var=equal_var)\n",
    "\n",
    "    c = a - 1\n",
    "    e11 = 0.0\n",
    "    e21 = 0.0\n",
    "    e12 = 0.0\n",
    "    e22 = 0.0\n",
    "    while c != 0:\n",
    "        if samples[chr][g][c]['type'] == 'exon':\n",
    "            e11 = samples[chr][g][c]['TPM_ctrl']\n",
    "            e21 = samples[chr][g][c]['TPM_treat']\n",
    "            break\n",
    "        c -= 1\n",
    "    c = a + 1\n",
    "    while c != len(samples[chr][g]):\n",
    "        if samples[chr][g][c]['type'] == 'exon':\n",
    "            e12 = samples[chr][g][c]['TPM_ctrl']\n",
    "            e22 = samples[chr][g][c]['TPM_treat']\n",
    "            break\n",
    "        c += 1\n",
    "    samples[chr][g][a]['p'] = p\n",
    "    minuslog10p = -1.0 * math.log10(p)\n",
    "    r1 = math.log2(samples[chr][g][a]['TPM_ctrl'] / (e11 + e12))\n",
    "    r2 = math.log2(samples[chr][g][a]['TPM_treat'] / (e21 + e22))\n",
    "    log2TPMRatio = math.log2(samples[chr][g][a]['TPM_treat'] / samples[chr][g][a]['TPM_ctrl'])  \n",
    "    return p, minuslog10p, r1, r2, log2TPMRatio"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Function to process all data\n",
    "\n",
    "This function process all genes with all introns. The results are inserted into the \"samples\" dictionary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_sample(samples):   \n",
    "    count = 0\n",
    "    for chr in samples:\n",
    "        for g in samples[chr]:\n",
    "            for a in samples[chr][g]:\n",
    "                try:\n",
    "                    if samples[chr][g][a]['type'] == 'intron':\n",
    "                        p, minuslog10p, r1, r2, log2TPMRatio = process_one(chr, g, a, samples)\n",
    "                        samples[chr][g][a]['p'] = p\n",
    "                        samples[chr][g][a]['minuslog10p'] = minuslog10p\n",
    "                        samples[chr][g][a]['r1'] = r1\n",
    "                        samples[chr][g][a]['r2'] = r2\n",
    "                        samples[chr][g][a]['log2TPMRatio'] = log2TPMRatio                   \n",
    "                except Exception as inst:\n",
    "                    pass\n",
    "    return samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_data = process_sample(samples_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Function to create the arrays to be ploted with colors\n",
    "\n",
    "This function return tree arrays x, y, c. The arrays x and y are the points to be plotted. x is $log_{2}\\lgroup\\frac{TPM^{intron}_{treat}}{TPM^{intron}_{ctrl}}\\rgroup$ and y is $-1.0 * log_{10}(PValue)$. c is an array of colors with red for statistically relevant points and black for the non relevants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_plot(samples, r, p):\n",
    "    x_intron = []\n",
    "    y_intron = []\n",
    "    c_intron = []\n",
    "    for chr in samples:\n",
    "        for g in samples[chr]:\n",
    "            for a in samples[chr][g]:\n",
    "                if samples[chr][g][a]['type'] == 'intron' and 'r1' in samples[chr][g][a]:\n",
    "                    if samples[chr][g][a]['r1'] > r or samples[chr][g][a]['r2'] > r:\n",
    "                        y_intron.append(samples[chr][g][a]['minuslog10p'])\n",
    "                        x_intron.append(samples[chr][g][a]['log2TPMRatio'])\n",
    "                        if abs(samples[chr][g][a]['log2TPMRatio']) > 2 and samples[chr][g][a]['p'] < p:\n",
    "                            c_intron.append('red')\n",
    "                        else:\n",
    "                            c_intron.append('black')                    \n",
    "    return x_intron, y_intron, c_intron    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_intron, y_intron, c_intron = process_plot(samples_data, -100.0, 0.005)\n",
    "plt.xlabel(r'$log_{2}(\\frac{TPM_{Intron/cond2}}{TPM_{Intron/cond1}})$')\n",
    "plt.scatter(x_intron, y_intron, s=1, c=c_intron)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Printing the selected data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_process_plot(samples, r, p, filename):\n",
    "    count = 0\n",
    "    with open(filename, 'w') as f_out:\n",
    "        for chr in samples:\n",
    "            for g in samples[chr]:\n",
    "                for a in samples[chr][g]:\n",
    "                    if samples[chr][g][a]['type'] == 'intron' and 'minuslog10p' in samples[chr][g][a]:\n",
    "                        if abs(samples[chr][g][a]['log2TPMRatio']) > 2 and samples[chr][g][a]['p'] < p and (samples[chr][g][a]['r1'] > r or samples[chr][g][a]['r2'] > r):\n",
    "                            count += 1\n",
    "                            f_out.write(\n",
    "                                g + '\\t' \n",
    "                                + chr + ':' + str(samples[chr][g][a]['start']) + '-' + str(samples[chr][g][a]['end'])\n",
    "                                + '\\t' + str(samples[chr][g][a]['p'])\n",
    "                                + '\\t' + str(samples[chr][g][a]['minuslog10p'])\n",
    "                                + '\\t' + str(samples[chr][g][a]['TPM_ctrl'])\n",
    "                                + '\\t' + str(samples[chr][g][a]['TPM_treat'])                            \n",
    "                                + '\\t' + str(samples[chr][g][a]['log2TPMRatio'])\n",
    "                                + '\\t' + str(samples[chr][g][a]['r1'])\n",
    "                                + '\\t' + str(samples[chr][g][a]['r2'])\n",
    "                                + '\\n')\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "count = print_process_plot(samples_data, -1.0, 1E-6, 'diff_exp_intron_retained.txt')\n",
    "print(count)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
